### PR TITLE
Donovan crop overhaul

### PR DESCRIPTION
**Pull request by wmaxdonovan**
Primarily updated commenting and organization + a few updates to make sure pseudocode, spreadsheet model, and RUFAS were up to date. None of these changes were intended to alter output and diagnostics show 100% overlap (still figuring out how to share diagnostics without sending 25+ .fig files). pseudocode attached. Spreadsheet changes hi lighted in yellow. 

[pseudocode_SC_actualgrowth.docx](https://github.com/RuminantFarmSystems/MASM/files/2993159/pseudocode_SC_actualgrowth.docx)
[pseudocode_SC_cropbiomass.docx](https://github.com/RuminantFarmSystems/MASM/files/2993160/pseudocode_SC_cropbiomass.docx)
[pseudocode_SC_cropyield.docx](https://github.com/RuminantFarmSystems/MASM/files/2993161/pseudocode_SC_cropyield.docx)
[pseudocode_SC_growthconstraints.docx](https://github.com/RuminantFarmSystems/MASM/files/2993162/pseudocode_SC_growthconstraints.docx)
[pseudocode_SC_hydrology.docx](https://github.com/RuminantFarmSystems/MASM/files/2993163/pseudocode_SC_hydrology.docx)
[pseudocode_SC_nitrogenuptake.docx](https://github.com/RuminantFarmSystems/MASM/files/2993164/pseudocode_SC_nitrogenuptake.docx)
[pseudocode_SC_phosphorusuptake.docx](https://github.com/RuminantFarmSystems/MASM/files/2993165/pseudocode_SC_phosphorusuptake.docx)
[pseudocode_SC_wateruptake.docx](https://github.com/RuminantFarmSystems/MASM/files/2993167/pseudocode_SC_wateruptake.docx)

### CODE REVIEW

File: RUFAS/routines/crop/yields.py [hunk 1]
```diff
@@ -149,9 +149,10 @@ def calc_nutrient_removal(crop_type):
...
 #
 # Updates the current residue.
-# "Pseudo code_SC_actual growth and yield_1.0.docx" section 7.B.4
+# "pseudocode_SC_actualgrowth.docx" section 7.B.4
 #
 def calc_residue(crop_type, time):
     if crop_type.harvest_date == time.day:
-        dResidue = crop_type.biomass_actual - crop_type.yield_actual
+        # dResidue = yld * (1 - harv_eff) = yld - yld_actual = (bio_AF * HI) - yld_actual
+        dResidue = (crop_type.bio_AG * crop_type.HI_max) - crop_type.yield_actual
```
Comment (mmotew) [inline RUFAS/routines/crop/yields.py]: I'm having trouble following this correction on line 157. I've tried to trace back the calculation in the spreadsheet model with no luck. Conceptually the original equation here makes sense. Another thing to keep in mind is that we may not be growing crops right now (you should double check) so you may not see anything in the Rufas output to flag a difference.

File: RUFAS/routines/crop/soil_water_uptake.py [hunk 2]
```diff
@@ -127,20 +128,20 @@ def inc_lower_layer_uptake(crop_type, soil, uptake_each_layer):
...
+# "pseudocode_SC_wateruptake.docx" section 2.C.2
 #
 def decrease_effic_of_uptake(soil, uptake_each_layer):
     adjusted_uptakes = []
 
     for uptake, layer in zip(uptake_each_layer, soil.listOfSoilLayers):
         # Point at which plant available water in soil layer begins to limit
         # efficiency of plant uptake (mm H2O)
-        AWC = 0.25 * (layer.fcWater - layer.wiltingWater) + layer.wiltingWater
+        AWC_limit = 0.25 * (layer.fcWater - layer.wiltingWater) + layer.wiltingWater
```
Comment (mmotew) [inline RUFAS/routines/crop/soil_water_uptake.py:139]: I take it that this new variable name is simply more intuitive?

File: RUFAS/routines/crop/leaf_area_index.py [hunk 3]
```diff
@@ -108,17 +107,33 @@ def calculate_LAI_actual(crop_type, time):
...
+
+        # 1.D.4.a.1
         crop_type.LAI_actual = crop_type.prev_LAI_actual + dLAI_actual
 
     else:
-        result = crop_type.LAI_max * (1 - crop_type.fr_PHU) / (1 - crop_type.fr_PHU_sen)
-        crop_type.LAI_actual = max(result, 0)
+
+        # 1.D.4.a.2
+        LAI_actual = crop_type.LAI_max * (1 - crop_type.fr_PHU) / (1 - crop_type.fr_PHU_sen)
```
Comment (mmotew) [inline RUFAS/routines/crop/leaf_area_index.py:122]: I'm assuming you looked at LAI_actual with the diagnostic script? Just verifying there's no change.

**Review by mmotew — COMMENTED** [review]
Being able to see those diagnostics will help me verify that the code changes are not changing model behavior. Could you maybe share a folder with the figures, saved in a compressed format? 

Can you verify that Rufas is running a crop? Output comparisons between changed code vs old code could be misleading if you aren't actually simulating a crop. (The spreadsheet model isn't running a crop right now, which is why I'm mentioning this.)

And along those lines, I wonder if there is a way to include some basic simulation specs to the code reviewers so we can see how you set up the run. I think at some point this will be important. Just something to think about. 

Checking every variable affected by a code change is in general a good idea, even the little helper variables. Maybe your IDE can let you see or output whole vectors of the variable directly so you don't have to monkey with I/O routines. (This is another reason I like to write models in MATLAB, although that won't help us here...)

File: RUFAS/routines/crop/crop.py [hunk 4]
```diff
@@ -129,57 +125,43 @@ def daily_crop_routine(crop, weather, time, soil):
         #
         heat_units.update_all(crop_type, T_min, T_max, time)
 
+        leaf_area_index.update_all(crop_type, time)
```
Comment (mmotew) [inline RUFAS/routines/crop/crop.py]: Changing the order of updates can affect model output in unpredictable ways. Biomass and LAI are linked, for example, so order may matter a lot. Can you verify that model output is unaffected by these changes?

File: RUFAS/routines/crop/biomass.py [hunk 5]
```diff
@@ -71,23 +71,23 @@
 # This function updates all biomass information
 #
 def update_all(crop_type, time, weather, soil):
-    # update gamma_reg value
-    calc_gamma_reg(crop_type, time, weather, soil)
 
     # update biomass values
     calc_actual_Biomass(crop_type, time, weather)
 
+    # update gamma_reg value
```
Comment (mmotew) [inline RUFAS/routines/crop/biomass.py]: Again has the order of operations here changed? Could it matter? (i.e. does gamma_reg affect biomass?)

File: RUFAS/routines/crop/biomass.py [hunk 6]
```diff
@@ -134,89 +134,91 @@ def calc_gamma_reg(crop_type, time, weather, soil):
...
     phi_p = calc_phi_P(crop_type)
-    result = 1 - phi_p / (phi_p + exp(3.535 - 0.02597 * phi_p))
-    return min(0.99, result)
+    pstrs = 1 - phi_p / (phi_p + exp(3.535 - 0.02597 * phi_p))
+    return min(0.99, pstrs)
 
 
 #
 # Calculates phosphorus stress scaling factor.
-# "Pseudo code_SC_growth constraints_1.0.docx" section 6.4.1
```
Comment (mmotew) [inline RUFAS/routines/crop/biomass.py:216]: I think there's a small bug in the pseudocode file for the phosphorus stress variable. It's named "nstrs" and should really be "pstrs". Somebody just copied and pasted I think.

File: RUFAS/routines/crop/biomass.py [hunk 7]
```diff
@@ -134,89 +134,91 @@ def calc_gamma_reg(crop_type, time, weather, soil):
...
+# "pseudocode_SC_growthconstraints.docx" section 6.4.1
 #
 def calc_phi_P(crop_type):
     if crop_type.bio_P_opt == 0:
         return 300
     else:
-        return 200 * ((crop_type.bio_P/crop_type.bio_P_opt) - 0.5)
+        phi_p = 200 * ((crop_type.bio_P/crop_type.bio_P_opt) - 0.5)
+        return max(0, phi_p)
```
Comment (mmotew) [inline RUFAS/routines/crop/biomass.py]: I don't see anything in the spreadsheet model or pseudocode for max(0, phi_p). Did you find a reason this variable needs to be above 0?

File: RUFAS/routines/crop/yields.py [hunk 8]
```diff
@@ -153,6 +153,6 @@ def calc_nutrient_removal(crop_type):
 #
 def calc_residue(crop_type, time):
     if crop_type.harvest_date == time.day:
-        # dResidue = yld * (1 - harv_eff) = yld - yld_actual = (bio_AF * HI) - yld_actual
-        dResidue = (crop_type.bio_AG * crop_type.HI_max) - crop_type.yield_actual
+        # matches spreadsheet model, but not pseudocode
+        dResidue = crop_type.biomass_actual - crop_type.yield_actual
```
Comment (mmotew) [inline RUFAS/routines/crop/yields.py:157]: Ok, it looks like you went back and changed this to what it was previously, and it makes sense to me that way. I think you've updated this in the pseudocode too (highlighted) but I'd be interested to see what the equation was previously.

File: RUFAS/routines/crop/biomass.py [hunk 9]
```diff
@@ -72,11 +72,12 @@
 #
 def update_all(crop_type, time, weather, soil):
 
+    # update gamma_reg value
```
Comment (mmotew) [inline RUFAS/routines/crop/biomass.py:75]: Did you have to change the order of operations here back to what it was?

File: RUFAS/routines/crop/biomass.py [hunk 10]
```diff
@@ -219,6 +220,5 @@ def calc_phi_P(crop_type):
     if crop_type.bio_P_opt == 0:
         return 300
     else:
-        phi_p = 200 * ((crop_type.bio_P/crop_type.bio_P_opt) - 0.5)
-        return max(0, phi_p)
```
Comment (mmotew) [inline RUFAS/routines/crop/biomass.py]: Was the capping at 0 causing a problem? I'd be curious to know where the idea for it originated.

**Review by mmotew — COMMENTED** [review]
Another general comment: It looks like you may have made some commits early on that you later went back and changed (which is totally fine of course). It might be good to point that out to a code reviewer ahead of time so they can skip any changes that get overwritten. I realize that may be hard to do, depending on how much code is involved. So, just a thought.

**Comment by mmotew** [comment]
Hi Max,

I've left my review comments either as file-specific comments or line-specific comments. Hopefully you can see all of them. 

In general things looks good although I spent a bit of time trying to trace down some reasons for code changes that later on got revoked (I think). I'd still like to see the comparison plots, too. We can try that next week. 

I like the attention to formatting and general structure! Keeps things clean, organized and understandable for the next person.  

Missy

**Comment by mmotew** [comment]
Oops - I am reopening this in case Militsa hasn't submitted her review yet.

**Comment by militsasotirova** [comment]
This all looks reasonable to me! I think the updates to the comments look good as well.

Militsa

**Comment by wmaxdonovan** [comment]
[crop_overhaul_diagnostics 2.zip](https://github.com/RuminantFarmSystems/MASM/files/3009232/crop_overhaul_diagnostics.2.zip)

**Pull request merged**